tools/depcheck/pkg/cmd/trace.go

```diff
@@ -0,0 +1,299 @@
+
+	return cmd
+}
+
+func (o *TraceImportsOpts) Validate() error {
+	if o.YamlFile == nil || o.LockFile == nil {
+		return fmt.Errorf("both a glide.yaml and a glide.lock file are required")
+	}
+
+	return nil
+}
+
+type Node struct {
+	concrete.Node
+	uniqueName string
+	baseName string
+}
+
+type PackageImports struct {
+	Imports []string
+}
+
+func (o *TraceImportsOpts) Run() error {
+	g, nodesByRepo, err := o.buildGraph()
+	if err != nil {
+		return err
+	}
+
+	// if no root import given, calculate all unique glide.yaml imports
+	if len(o.RootImport) == 0 {
+		deps, err := o.uniqueNonTransitiveDeps(g, nodesByRepo)
+		if err != nil {
+			return err
+		}
+
+		fmt.Printf("%v\n", deps)
+		return nil
+	}
+
+
+	return o.uniqueTransitiveDeps()
+}
+
+func (o *TraceImportsOpts) buildGraph() (*concrete.DirectedGraph, map[string][]Node, error) {
+	g := concrete.NewDirectedGraph()
+	nodeMap := make(map[string]Node)
+	nodesByRepoName := make(map[string][]Node)
+
+	for _, repo := range o.LockFile.Imports {
+		repoName := repo.Name
+		if len(o.BaseDir) > 0 {
+			repoName = strings.Join([]string{o.BaseDir, repoName}, "/")
+		}
+
+		nodesByRepoName[repoName] = []Node{}
+
+		if len(repo.SubPackages) == 0 {
+			n := Node{concrete.Node(g.NewNodeID()), repoName, repoName}
+			nodeMap[n.uniqueName] = n
+			nodesByRepoName[repoName] = append(nodesByRepoName[repoName], n)
+			continue
+		}
+
+		for _, pkg := range repo.SubPackages {
+			n := Node{
+					Node: concrete.Node(g.NewNodeID()),
+					uniqueName: strings.Join([]string{repoName, pkg}, "/"),
+					baseName: repoName,
+				}
+			nodeMap[n.uniqueName] = n
+			nodesByRepoName[repoName] = append(nodesByRepoName[repoName], n)
+		}
+	}
+
+	errs := []error{}
+	for _, n := range nodeMap {
+		if err := addToGraph(g, n, nodeMap); err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	if len(errs) > 0 {
+		aggregate := "The following errors ocurred while populating the dependency graph:\n"
+		for _, e := range errs {
+			aggregate += fmt.Sprintf("%v\n", e)
+		}
+
+		return nil, nodesByRepoName, fmt.Errorf("%v", aggregate)
+	}
+
+	return g, nodesByRepoName, nil
+}
+
+func (o *TraceImportsOpts) uniqueNonTransitiveDeps(g *concrete.DirectedGraph, nodesByRepo map[string][]Node) ([]string, error) {
+	unique := []string{}
+	for _, repo := range o.YamlFile.Imports {
+		repoName := repo.Package
+		if len(o.BaseDir) > 0 {
+			repoName = strings.Join([]string{o.BaseDir, repoName}, "/")
+		}
+
+		nodes, exists := nodesByRepo[repoName]
+		if !exists {
+			fmt.Fprintf(o.ErrOut, "info: non-transitive dependency %q not found in glide.lock imports", repoName)
+			unique = append(unique, repoName)
+			continue
+		}
+
+		if !isNodeSetUnique(g, nodes) {
+			continue
+		}
+
+		unique = append(unique, repoName)
+	}
+
+	return unique, nil
+}
+
+func (o *TraceImportsOpts) uniqueTransitiveDeps() error {
+	// TODO
+	return nil
+}
+
+func addToGraph(g *concrete.DirectedGraph, n Node, nodes map[string]Node) error {
```

Comment:
Ok, I think I understand what this is trying to do.  Why wouldn't you build this graph based on the `go list` of a starting package or packages?  Seems like you could build the whole graph from that.  I might skip core libraries (there only nine different sources for our vendor deps).

After you built such a graph, you could build a second graph that had the "blame" packages (`github.com/openshift/origin/pkg/<folder>`) and every other repo as repos to make the graph much smaller.  Once you had that graph, you could start running traversals, right?

Expensive, but simple.